src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -183,56 +219,98 @@ protected override void Dispose(bool disposing)
-                    var rcs = (RequestCompletionSource)s;
-                    rcs.Handler.StartRequest(rcs);
-                },
-                state,
-                CancellationToken.None,
-                TaskCreationOptions.DenyChildAttach,
-                TaskScheduler.Default);
+            BeginRequest(state);
 
             return state.Task;
         }
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void BeginRequest(RequestCompletionSource state)
         {
-            HttpResponseMessage responseMessage = null;
-            SafeCurlHandle requestHandle = null;
+            SafeCurlHandle requestHandle = new SafeCurlHandle();
+            GCHandle stateHandle = new GCHandle();
+            bool needCleanup = false;
 
-            if (state.CancellationToken.IsCancellationRequested)
+            try
             {
-                state.TrySetCanceled();
-                return;
-            }
+                // Prepare context objects
+                state.ResponseMessage = new CurlResponseMessage(state.RequestMessage);
+                stateHandle = GCHandle.Alloc(state);
+                requestHandle = CreateRequestHandle(state, stateHandle);
+                state.RequestHandle = requestHandle;
+                needCleanup = true;
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+                if (state.CancellationToken.IsCancellationRequested)
+                {
+                    state.TrySetCanceled(state.CancellationToken);
+                    return;
+                }
 
-            try
-            {
-                requestHandle = CreateRequestHandle(state);
+                if (state.RequestMessage.Content != null)
+                {
+                    Stream requestContentStream =
+                        await state.RequestMessage.Content.ReadAsStreamAsync().ConfigureAwait(false);
+                    if (state.CancellationToken.IsCancellationRequested)
+                    {
+                        state.TrySetCanceled(state.CancellationToken);
+                        return;
+                    }
+                    state.RequestContentStream = requestContentStream;
+                    state.RequestContentBuffer = new byte[s_requestBufferSize];
+                }
 
-                state.CancellationToken.ThrowIfCancellationRequested();
-                int result = Interop.libcurl.curl_easy_perform(requestHandle);
-                if (CURLcode.CURLE_OK != result)
+                AddEasyHandle(state);
+                needCleanup = false;
+            }
+            catch (Exception ex)
+            {
+                HandleAsyncException(state, ex);
+            }
+            finally
+            {
+                if (needCleanup)
+                {
+                    RemoveEasyHandle(_multiHandle, stateHandle, false);
```

Comment:
Yes... we still need to do the required freeing up like curl_easy_cleanup. Special care is required for the AddRef which is being done in the catch block of AddEasyHandle